### DISCUSSION TITLE
Error during cropping a large image in 7.1.1-33 version

### DISCUSSION THREAD

**Discussion by MishaPhysics** [ImageMagick/ImageMagick#7392] [Help]
### ImageMagick version

7.1.1-33

### Operating system, version and so on

Windows 10

### Description

I have a picture "file.jpg" of 32768x32768 resolution, and after running the command `magick mogrify -crop 512x512 file.jpg` I get the error:
`mogrify: unable to extend cache 'file.jpg': Invalid argument @ error/cache.c/OpenPixelCache/3925.`
It happens with "ImageMagick-7.1.1-33-Q16-HDRI-x64-dll".
It works fine (without the error) with an old "ImageMagick-7.1.1-14-Q16-HDRI-x64-dll".

Also, it works fine with both IM versions if the image is 16384x16384.

**Comment by fmw42** [comment]
With your command you are trying to create 4096 512x512 images.   Do you really want that?  You cannot create multiple outputs for magick mogrify.  If you really want to do that and have enough RAM, use `magick file.jpg -crop 512x512 file_%04d.jpg` That will number your files with leading zeros.

**Comment by snibgo** [comment]
> magick mogrify -crop 512x512 file.jpg

As a general rule: if you have only one input image, don't use `magick mogrify`. Instead, use just `magick`.

`magick mogrify` is used when you have multiple inputs, and you want to apply the same processing to each of them.

**Reply by MishaPhysics** [reply]
[fmw42](https://github.com/fmw42), yes, I want to get 4096 512x512 images.
The command `magick file.jpg -crop 512x512 file_%04d.jpg` gives the same error.

[snibgo](https://github.com/snibgo), the command `magick -crop 512x512 file.jpg` doesn't work and give an error, whereas `magick mogrify -crop 512x512 file.jpg` works perfectly and crops my single input picture (but if a picture is 16384x16384, not 32768x32768).

The command `magick mogrify -crop 512x512 file.jpg` works perfectly for me with the old "ImageMagick-7.1.1-14-Q16-HDRI-x64-dll". I've checked this many times unistalling  "ImageMagick-7.1.1-33-Q16-HDRI-x64-dll" and installing "ImageMagick-7.1.1-14-Q16-HDRI-x64-dll". The problem seems to be with memory, since the command `magick mogrify -crop 512x512 file.jpg` works perfectly even with IM 7.1.1-33 if my input file is 16384x16384. So, maybe in a new IM version it was made some changes regarding this.

**Reply by fmw42** [reply]
You are running out of RAM or cache.  Or perhaps you have run into your limit on input file width. Change your policy.xml file.

**Reply by MishaPhysics** [reply]
There are two policy.xml files. The one is in the main IM directory, and the second is in the \www\source subdirectory. Which one should I change? Could you tell me which parameter I should increase?

**Reply by fmw42** [reply]
What do you mean for "main IM directory"?  

I set mine at /usr/local/etc/ImageMagick-7/policy.xml

**Comment by fmw42** [comment]
From https://imagemagick.org/script/architecture.php#tera-pixel

For large images, memory resources will likely be exhausted and ImageMagick will instead create a pixel cache on disk. Make sure you have plenty of temporary disk space. If your default temporary disk partition is too small, tell ImageMagick to use another partition with plenty of free space. For example:

```
magick -define registry:temporary-path=/data/tmp logo:  \ 
     -resize 250000x250000 logo.miff
```

To ensure large images do not consume all the memory on your system, force the image pixels to memory-mapped disk with resource limits:

```
magick -define registry:temporary-path=/data/tmp -limit memory 16mb \
  logo: -resize 250000x250000 logo.miff
```

Here we force all image pixels to disk:

```
magick -define registry:temporary-path=/data/tmp -limit area 0 \
  logo: -resize 250000x250000 logo.miff
```

Caching pixels to disk is about 1000 times slower than memory. Expect long run times when processing large images on disk with ImageMagick. You can monitor progress with this command:

```
magick -monitor -limit memory 2GiB -limit map 4GiB -define registry:temporary-path=/data/tmp \
  logo: -resize 250000x250000 logo.miff
```

**Reply by fmw42** [reply]
What do you have now for 

`magick -list resource`

Is your image color or grayscale?  And does it have an alpha channel.

I am not an expert on this but, If your image is 32768\*32768 3 channels at 8-bits per channel, then it needs 3221225472 bytes = 3.2 Gigabytes. You would then need double that for input and output, so 6.4 Gigabytes.  I presume your RAM is more than than.  If so, be sure in your resource via your policy that you have enough memory allocated then increase your Map size and be sure your disk is unlimited or larger enough to hold your 6.4 gigabytes.  Perhaps one of the other users might comment or correct this.

**Reply by fmw42** [reply]
See also https://imagemagick.org/script/security-policy.php

**Reply by MishaPhysics** [reply]
`magick -list resource` shows:
![im](https://github.com/ImageMagick/ImageMagick/assets/104359915/b67d8b93-f1a0-4857-b7fd-fabb36fc62c3)

**Reply by MishaPhysics** [reply]
That was for IM 7.1.1-33.

For the IM 7.1.1-14 I get:
![im2](https://github.com/ImageMagick/ImageMagick/assets/104359915/38e720fd-29af-44db-8278-075b8137478e)
This must be the reason.

**Reply by fmw42** [reply]
Your memory is marginal so is probably running out.  How much RAM do you have on your OS?  If larger, set the memory in your policy.xml file bigger.  If your RAM is no larger than the memory in your policy.xml, then increase your MAP.  Not sure is you can have more MAP than your RAM.  So I presume you might have 16 GiB of RAM. If so and if the MAP cannot be larger, then increasing it might not work.

Perhaps one of the more expert users (@snibgo or a developer) can comment further about Map limitations vs actual RAM or if MAP is coming from disk and so not as limited.  I do not know this that well.

**Reply by MishaPhysics** [reply]
Yes, I have 16 Gb of RAM.

**Reply by fmw42** [reply]
Then try setting your Memory to say 10 GiB. If that fails try 12 or 14 or 16. If that fails, try increasing your Map to say 20 GiB. Be sure to check that the change was actually made and show in your `magick -list resource`.

**Comment by snibgo** [comment]
32768x32768 pixels is 1 GP. At 3 channels, and 4 bytes per channel, the input image needs 12 GB of memory. The same again for the output, so that is 24 GB, which is more than your available memory. Memory is also needed for other purposes, and perhaps all your memory has been used for the pixels, and is not available when needed for the other purposes.

So I suggest *limiting* the memory (and map) that are used for pixels, so not as much memory is used for the pixels. For example:
```
magick -limit memory 1GB -limit map 2GB file.jpg -crop 512x512 out_%06d.jpg
```
Ensure you have plenty of free space on your disk.

**Reply by fmw42** [reply]
Thanks snibgo for the correction.

**Reply by fmw42** [reply]
> 32768x32768 pixels is 1 GP. At 3 channels, and 4 bytes per channel, the input image needs 12 GB of memory

Sorry, I am not understanding where the 4 bytes per channel come from for 3 channels.  We need 3 or 4 bytes per pixel to account for either RGB or RGBA channels -- that is 1 Byte per channel (or 8-bits per channel).  So we have for 3 channels, 1 GP x 3 bytes = 3 Gibytes.   What am I mis-construing?

**Comment by snibgo** [comment]
The build of IM is Q16, with HDRI. The OP says `ImageMagick-7.1.1-33-Q16-HDRI-x64-dll` and
`ImageMagick-7.1.1-14-Q16-HDRI-x64-dll`'. So each sample is recorded in memory as a 32-bit float, so 4 bytes per channel. If the image has 3 channels (R, G and B) then it needs 12 bytes per pixel.

As the image is JPEG, the version of IM used could be Q8, without HDRI. That would need only 1 byte per channel, so 3 bytes per pixel.

The program `jpegtran` can be used to chop a JPEG into tiles. It might need far less memory, as it doesn't need to de-compress the JPEG input, and re-compress for the output.

**Comment by MishaPhysics** [comment]
I ran exactly your command `magick -limit memory 1GB -limit map 2GB file.jpg -crop 512x512 out_%06d.jpg`, and got the error:
`magick: unable to extend cache 'file.jpg': Invalid argument @ error/cache.c/OpenPixelCache/3925.`
But it's not so problem. I could just install the ImageMagick-7.1.1-14 if I want to crop a picture which is 32768x32768. It works fine with that version.
